### ISSUE TITLE
Enable client interceptors for additional metadata

### ISSUE THREAD

**Issue by crisbodnar** [opentracing-contrib/java-grpc#15]
I am trying to write additional Interceptors which add more metadata (project specific) to the spans created by the library interceptors. While this is possible for server spans, I believe it is not possible for client spans. The following code would need to be changed:

```
private Span createSpanFromParent(SpanContext parentSpanContext, String operationName) {
    if (parentSpanContext == null) {
      return tracer.buildSpan(operationName).start();
    } else {
      return tracer.buildSpan(operationName).asChildOf(parentSpanContext).start();
    }
  }
```
to
```
private Span createSpanFromParent(SpanContext parentSpanContext, String operationName) {
    if (parentSpanContext == null) {
      return tracer.buildSpan(operationName).startActive(false);
    } else {
      return tracer.buildSpan(operationName).asChildOf(parentSpanContext).startActive(false);
    }
  }
```

Then the client interceptor would be able to get the span from the current scope. 
Please let me know if there is any other way of doing this which I might not be aware of.

**Comment by malafeev** [comment]
@crisbodnar `startActive(false)` will create active scope which should be closed, otherwise we will get a leak.
Maybe better solution would be add additional SpanDecorator interface to `ClientTracingInterceptor`.
You will be able to provide own implementation of `SpanDecorator` and in `ClientTracingInterceptor.interceptCall(...)` `SpanDecorator` will add additional metadata to span.

```
interface ClientSpanDecorator {
 
    void interceptCall(Span span, MethodDescriptor<ReqT, RespT> method,
      CallOptions callOptions);

}
```
cc @tedsuo

**Comment by crisbodnar** [comment]
The SpanDecorater interface sounds very good @malafeev ! I think the need to custom decorate the span is quite common. For example, I find the current way of putting all headers in a single tag quite problematic because in the backend UI I can't search for gRPC calls with a specific header value. Another reason we were doing this was to address the issue and put each header in a separate tag.

The main reason is that we want to filter traces in the backend UI by specific deployments and this information is [Improbable](https://improbable.io/) specific. This solution would make the interceptors very customizable in this regard.

Would be good to have something like this for ServerTracingInterceptor as well to avoid the need for one additional interceptor to decorate the span. Let me know if we can proceed with this solution. I am also happy to implement it and send a PR ASAP.

**Comment by malafeev** [comment]
@crisbodnar sounds good. You are welcome to submit PR.

**Comment by crisbodnar** [comment]
Would you mind If I send a separate PR to add one tag per header rather than one tag for all headers or is there a particular reason it is implemented like this right now @malafeev ?

**Comment by malafeev** [comment]
In the PR you need to add only interface `ClientSpanDecorator` which should be called inside `ClientTracingInterceptor.interceptCall(...)`. If it is set of course.
So you need to add setter in  `ClientTracingInterceptor.Builder`.
Which tags will be added to span depends on decorator implementation.
Any user can define own implementation of decorator. We provide only interface.

**Comment by crisbodnar** [comment]
Yes. I know that. What I meant was referring to the other issue (not the SpanDecorator interface) of putting all headers in a single tag:
```
span.setTag("grpc.headers", headers.toString());
```
It would be good if this would put by default all headers in separate tags. This would be a separate Pull Request as it is a separate issue.

**Comment by crisbodnar** [comment]
Ok. Cool. Will start working on the SpanDecorator.

**Comment by crisbodnar** [comment]
@malafeev I am done with the PR. Could you give me write access to the repo so I can submit the PR?

**Comment by malafeev** [comment]
@crisbodnar you don't need write access to submit PR.

**Comment by malafeev** [comment]
you can do it from your fork

**Comment by crisbodnar** [comment]
Submitted the PR: https://github.com/opentracing-contrib/java-grpc/pull/16

**Comment by malafeev** [comment]
fixed. will be available in version 0.0.7 (in next 12 hours)
